enhancement(aws_cloudwatch_logs sink): Enforce age requirements

ktff opened the pull request:
Closes #1483 

### Open questions
 - [x] Regarding 

    > or older than the retention period of the log group

    part of the second constraint I'm unsure in how to deal with it? 
    Actually implementing this would somewhat complicate the sink, for a case which would happen rarely. If this isn't added, in the worst case `aws` may drop that batch which would have a limited effect since we are splitting the batches into 24h pieces. And dropping older messages also shouldn't be a problem since there is already a precedent of dropping messages older than 14 days. (EDIT: filtering on 14day should be enough) 

<!--
**Your PR title must conform to the conventional commit spec!**

  <type>!?(<scope>): <description>

  * `type` = chore, docs, enhancement, newfeat, perf
  * `!` = signals a breaking change
  * `scope` = https://github.com/timberio/vector/blob/master/.github/semantic.yml#L4
  * `description` = short description of the change

Examples:

  * enhancement(file source): Added `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fixed a bug discovering new files
  * perf(observability): Improved logging performance
  * docs: Clarified `batch_size` option
-->

binarylogic commented on src/sinks/aws_cloudwatch_logs/request.rs:
```diff
@@ -169,15 +173,22 @@ impl Future for CloudwatchFuture {
 
-                    info!(message = "putting logs was successful.", ?next_token);
+                    if let Some(events) = self.events.take() {
+                        info!(message = "putting logs.", ?next_token);
```
Is this necessary? Or should it be moved to `debug`? Don't we already have request logs?

ktff replied:
> Don't we already have request logs?

We have, but this will be called only in rare scenarios so I would like to make it clear when it happens, and I agree that `info` is too strong for this, `debug` would better fit it's purpose.

binarylogic commented:
Thanks @ktff. Just confirming, performance-wise does anything jump out here that might reduce that?

ktff commented:
@binarylogic The `sort` we introduced in #2320 , which I just moved in this PR to another location, it could on larger batches. It helps that in most cases messages will be almost sorted.

Other than that, there is an additional passage over messages for splitting them into 24h batches, but I now see that we can optimize it and introduce a fast path for majority case.

binarylogic commented:
Perfect, thanks!

LucioFranco commented on src/sinks/aws_cloudwatch_logs/mod.rs:
```diff
@@ -882,6 +899,86 @@ mod integration_tests {
+            .map(|e| e.message.unwrap())
+            .collect::<Vec<_>>();
+
+        assert_eq!(output_lines, lines);
```
we should probably add a test that shows the output events to be less than what were sent to the sink?

ktff replied:
This does test that. The `lines` only contains that what should pass through the filters.

LucioFranco replied:
Does this also check separate batches are produced?

ktff replied:
It doesn't.

LucioFranco commented on src/sinks/aws_cloudwatch_logs/request.rs:
```diff
@@ -189,9 +200,39 @@ impl Client {
+        RusotoFuture<PutLogEventsResponse, PutLogEventsError>,
+        Option<Vec<InputLogEvent>>,
+    ) {
+        // We will only send events that happened in the 24h window since the oldest event.
```
What is the reason for handling this logic here and not doing it all combined in the outer module?

ktff replied:
At the time, it seemed like it would require the least amount of changes in this place. But, yes, we could handle this by pulling it back to the outer module and passing `Vec<Vec<InputLogEvent>>` to the future.

ktff commented on src/sinks/aws_cloudwatch_logs/mod.rs:
```diff
@@ -314,16 +315,29 @@ impl Service<Vec<Event>> for CloudwatchLogsSvc {
+            // We need to take into account the flight time from here to aws servers,
+            // so we are reducing timestamp acceptable range from both sides
+            // for this amount.
+            let buffer_time = Duration::minutes(1);
```
No, just a speculation and defensive programming on my side since I'm not sure how strictly does aws check for this.  

> two hours is the buffer 

Now I see, yes this is not needed, I'll remove it.

ktff commented on src/sinks/aws_cloudwatch_logs/mod.rs:
```diff
@@ -314,16 +315,29 @@ impl Service<Vec<Event>> for CloudwatchLogsSvc {
                 .collect::<Vec<_>>();
 
+            // Sort by timestamp
+            events.sort_by_key(|e| e.timestamp);
```
True. I missed this even though the same optimization was applied to the 24/h filter.

ktff replied:
After working on this for a bit, I don't think it's worth it. We can't avoid iterating since we need to transform `Event` to `InputLogEvent`. While yes we could almost completely avoid cost of the filter, it's just 2 integer comparisons and `&&` for each event, we would need to add additional complexity which would have a negative performance for smaller event counts, while when we do have something to filter out from the start of the `Vec` we would need to shift everything else. While the current solution, when we do have something to filter, has benefit of reducing number of events to be sorted. 

All in all, I would say it's a blurry tie on performance side, but simplicity is on the side of the current solution.

LucioFranco commented on src/sinks/aws_cloudwatch_logs/mod.rs:
```diff
@@ -314,27 +315,71 @@ impl Service<Vec<Event>> for CloudwatchLogsSvc {
+                        .unwrap_or_else(|at| at);
+
+                    // Can't be empty
+                    let remainder = events.drain(at..).collect();
```
do we want to use https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.split_off?

ktff replied:
Yes, yes we do.
